File: src/main/java/me/wolfyscript/customcrafting/recipes/data/IngredientData.java
/*
 *       ____ _  _ ____ ___ ____ _  _ ____ ____ ____ ____ ___ _ _  _ ____
 *       |    |  | [__   |  |  | |\/| |    |__/ |__| |___  |  | |\ | | __
 *       |___ |__| ___]  |  |__| |  | |___ |  \ |  | |     |  | | \| |__]
 *
 *       CustomCrafting Recipe creation and management tool for Minecraft
 *                      Copyright (C) 2021  WolfyScript
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package me.wolfyscript.customcrafting.recipes.data;

import com.wolfyscript.utilities.bukkit.world.items.CustomItem;
import me.wolfyscript.customcrafting.recipes.items.Ingredient;
import org.bukkit.inventory.ItemStack;

/**
 * Contains the data for a single Ingredient.
 * <p>
 * It contains:<br>
 * - the recipe slot this data belongs to,<br>
 * - the Ingredient of the recipe,<br>
 * - the CustomItem that was chosen from that Ingredient,<br>
 * - and the created ItemStack of the CustomItem.
 */
public record IngredientData(int recipeSlot, Ingredient ingredient, CustomItem customItem, ItemStack itemStack) {
}

File: core/api/src/main/kotlin/com/wolfyscript/customcrafting/core/recipes/data/IngredientData.kt
package com.wolfyscript.customcrafting.core.recipes.data

import com.wolfyscript.customcrafting.core.recipes.ingredient.Ingredient
import com.wolfyscript.scafall.items.ItemStackRef

/**
 * Holds information about an ingredient that was selected after a recipe was evaluated.
 */
interface IngredientData {

    /**
     * The slot of the ingredient in the inventory.
     * For example, the slot in the crafting grid
     */
    val invSlot: Int

    /**
     * The index of the ingredient in the recipe
     */
    val recipeIndex: Int

    /**
     * The ingredient associated with this information
     */
    val selectedIngredient: Ingredient

    val matchedItemStackRef: ItemStackRef

}